### ISSUE TITLE
inline loader not migrateable to webpack 5 when used in require.context

### ISSUE THREAD

**Issue by eps1lon** [webpack/webpack#12900]
# Bug report

**What is the current behavior?**

The documentation in https://webpack.js.org/guides/asset-modules/#replacing-inline-loader-syntax to replace inline loaders with Asset Modules does not apply to `require.context`. Specifically, it doesn't seem like I can exclude a source asset from being parsed by other loaders. This was possibly in webpack 4 with `require.context("!raw-loader!./code", false, /\.js$/);` but there seems to be no API for this behavior in webpack 5 i.e. 

```js
{
  test: /\.m?js$/,
  resourceQuery: /^(?!\?raw$).*/,
},
{
  resourceQuery: /raw/,
  type: "asset/source",
},
{
  test: /\.(js|jsx)$/,
  exclude: /node_modules/,
  use: ["babel-loader"],
},
```
does result in transpiled source assets.

**If the current behavior is a bug, please provide the steps to reproduce.**

1. Clone https://github.com/eps1lon/webpack-asset-modules-source-require-context
2. `yarn install`
3. `yarn start´
4. Open page

The source code is printed transpiled (notice the `/*#__PURE__*/React.createElement` transpilation artifact).

**What is the expected behavior?**

The source asset is not transpiled (or parsed by any other loader for that matter).

**Other relevant information:**
webpack version: 5.21.1
Node.js version: 12.20.0
Operating System: Ubuntu 20.04.2 LTS
Additional tools:

The issue seems similar to https://github.com/webpack/webpack/issues/5730?

**Comment by alexander-akait** [comment]
Intersect rules, please use:
```
module: {
    rules: [
      {
        type: "asset/source",
        resourceQuery: /raw/,
      },
      {
        test: /\.(js|jsx)$/,
        exclude: /node_modules/,
        resourceQuery: /^(?!\?raw$).*/,
        use: ["babel-loader"],
      },
    ],
  },
```

otherwise `test` is preferred

**Comment by eps1lon** [comment]
That's not what the documentation is saying though? Now I have to change any existing (and future) rule that might process assets loaded as source.

**Comment by alexander-akait** [comment]
@eps1lon You need do it only for rules where `test` can have intersections

**Comment by eps1lon** [comment]
> @eps1lon You need do it only for rules where `test` can have intersections

That's not adressing my problem though. I have to make sure anybody who adds rules considers this.

How does you comment adress the problem where I don't have access to the rule objects e.g. when extending a nextjs webpack configuration?

**Comment by alexander-akait** [comment]
> maybe we should privatize `resourceQuery` and `resourceFragment`

What about this? So you don't need exclude `resourceQuery` from other rules

**Comment by eps1lon** [comment]
> > maybe we should privatize `resourceQuery` and `resourceFragment`
> 
> What about this? So you don't need exclude `resourceQuery` from other rules

I'm not following. How would I then be able to migrate from inline-loaders? Based on file extension is not possible since I need some files both as regular JS and their raw source.

**Comment by alexander-akait** [comment]
Sorry, I don't understand you.

> How would I then be able to migrate from inline-loaders?

Using `resourceQuery`,

> Based on file extension is not possible since I need some files both as regular JS and their raw source.

`resourceQuery: /^(?!\?raw$).*/,`

I provide solution above.

Do you want to avoid using `resourceQuery: /^(?!\?raw$).*/,`? If yes, above I suggested that we could prioritize `resourceQuery` over `test`, so you don't need `resourceQuery: /^(?!\?raw$).*/,` in your config

**Comment by eps1lon** [comment]
>> How would I then be able to migrate from inline-loaders?

>Using resourceQuery,

But you said you want to "privatize ressourceQuery". I interpreted "privatize" as "not being part of the public API" i.e. not usable. 

Could you summarize how I would write the initial set of rules without changing the rule that applies the babel-loader?

**Comment by alexander-akait** [comment]
@eps1lon sorry, typo, I mean `prioritize`

> Could you summarize how I would write the initial set of rules without changing the rule that applies the babel-loader?

Right now only using `resourceQuery: /^(?!\?raw$).*/,`, we need think more about it

**Comment by sokra** [comment]
> resourceQuery: /^(?!\?raw$).*/

That's pretty un-readable. You could use `resourceQuery: { not: /raw/ }`.

But even better use `oneOf` rules:

``` js
rules: [
  {
    oneOf: [
      {
        resourceQuery: /raw/,
        type: "asset/source",
      },
      {
        test: /\.(js|jsx)$/,
        exclude: /node_modules/,
        use: ["babel-loader"],
      },
    ]
  }
],
```

write a helper function if this is too much nesting for you...

**Comment by eps1lon** [comment]
> That's pretty un-readable. You could use resourceQuery: { not: /raw/ }.

I just copied the docs :sweat_smile:: https://webpack.js.org/guides/asset-modules/#replacing-inline-loader-syntax

> But even better use oneOf rules:

That does not work for me. I do not have control over all the rules since I'm using nextjs. I could technically but the order of the rules isn't some public API. And even then the point about future rules still apply: I need to manually verify no rules conflicts whenever I update nextjs.

**Comment by alexander-akait** [comment]
@sokra using `oneOf` is not very convenient in many cases, also some developers can't have access to `webpack.config.js`

**Comment by sokra** [comment]
> I just copied the docs 😅

hmmm... https://github.com/webpack/webpack.js.org/pull/4734/files

**Comment by alexander-akait** [comment]
@sokra Some configuration have a lot of rules, using `oneOf` will be difficult, also most of our configurations will use `oneOf` because it is very popular

**Comment by sokra** [comment]
btw. there is another way, which would work without touching any other part of the config:

``` js
      {
        test: /\.txt$/,
        type: "asset/source",
      },
```

```
import source from "file.txt!=!./file.js";
```

This treats content of file.js as it was written in `file.txt`.

Not sure if this works with `require.context`, but you could try it with
`require.context("code.txt!=!./code", false, /\.js$/)`, but this might look a bit weird anyway.

Otherwise you could also go with:

```
import sources from "./all-code.js!";
```

and generate code like that in the `all-code.js` **loader**:

``` js
exports { default as abc } from "code/abc.js.txt!=!./code/abc.js";
exports { default as def } from "code/def.js.txt!=!./code/def.js";
// ...
```

**Comment by alexander-akait** [comment]
@sokra I think we should try to avoid webpack specific syntax in code, it will be problem in future...

**Comment by eps1lon** [comment]
> btw. there is another way, which would work without touching any other part of the config:

How would this look in require.context? Specifically:
```js
const requireRaw = require.context(
  '!raw-loader!../../src/pages/components/rating',
  false,
  /\.(js|md|tsx)$/,
);
```

**Comment by sokra** [comment]
> I think we should try to avoid webpack specific syntax in code, it will be problem in future...

Yeah that why I only mentioned the other solutions first.

But he is already using `require.context` as webpack-specific syntax and there is not official replacement for that and there probably won't be...

If you want a solution without any webpack magic you can go with:

1. Create a pre build script that collects all source code needed into a JSON file.
2. Import that JSON file during the build.

**Comment by slorber** [comment]
I have a distinct but similar case where I feel it's complicated to migrate to the asset modules system.

Docusaurus has a modular architecture with a plugin system. 

A Docusaurus image plugin should be able to "override" the default image asset module rule. 

It looks to me that before, the order mattered, and we just had to add the custom image loader at the beginning of the list rule to override the existing rule. So we have plugins that do:

```js
export default function docusaurusPlugin() {
  return {
    name: 'docusaurus-plugin-ideal-image',
    configureWebpack() {
      return {
        mergeStrategy: {
          'module.rules': 'prepend', // using webpack-merge prepend customization rule
        },
        module: {
          rules: [
            {
              test: /\.(png|jpe?g|gif)$/i,
              loader: "custom-image-loader",
            },
          ],
        },
      };
    },
  };
}
```

It seems that now, adding this rule at the beginning does not prevent the default asset loaders in Docusaurus core to be used, unless we use oneOf.

I don't feel like using oneOf is very good in this case because it would break our existing plugin ecosystem that relies on append/prepend in rules, and expose more implementation details of the core Webpack config to plugins that want to customize it. Playing with resourceQuery and notResourceQuery is also not very convenient in our case, as plugin authors would have to potentially add notResourceQuery clauses.

**Comment by cloudcoderuwu** [comment]
I'm not sure if this is the origin of my issue, but after upgrading to webpack 5, it seems impossible to access the actual content of the raw files.

I have tried both using `raw-loader`(legacy) and the new asset module approach, but when I import any file (in my case markdown) or use `require.context`, I am unable to call `.default` to get the text from the file, it returns undefined...

```
rules: [
          {
            resourceQuery: /\.md$/,
            type: "asset/source",
          },
```

```
const mdFilesRaw = require.context(
  'static/md/posts/', 
  true,
  /\.md$/ 
)
const mdContent = mdFilesRaw.keys().map( 
  path => (
    mdFilesRaw(path).default
  )
)
```

`mdFilesRaw(path)` will return a string that is a hashed version of the filename moved to the public folder, such as `/static/media/POST.##HASH##.md` but markdown is also not creating this file or anything...

**Comment by sokra** [comment]
Don't use `.default`?

**Comment by nwaxiomatic** [comment]
> Don't use `.default`?

How do you access the text in the file then? 
In this case I need to pass the markdown text (as text) to a markdown parser. 
Before Webpack 5 you would call `.default`

Currently just calling `mdFilesRaw(path)` returns a single string (as stated in my original comment), it does not return the text within the file.

**Comment by Airkro** [comment]
Is possible to support `type` as  `import.meta.webpackContext` parameter?

**Comment by vankop** [comment]
> Is possible to support type as import.meta.webpackContext parameter?

there is no `type` in `import.meta.webpackContext`. we don't plan to support more options for `require.context`.

**Comment by vankop** [comment]
there is an answer on original issue. please create separate issues.

**Comment by ITsvetkoFF** [comment]
For those who are struggling with

```
    {
       test: /\.m?js$/,
       resourceQuery: { not: [/raw/] },
       use: [ ... ]
     },
```
for a webpack config that you are not controlling completely (like in Storybook)

there is another approach that I (in a desperate moment) was able to find in [code search](https://github.com/search?q=resourceQuery%3A+%2Fraw%2F%2C+++++++++++type%3A+%27asset%2Fsource%27%2C+++++++++%7D%2C&type=code) over github: [source](https://github.com/DevCloudFE/ng-devui/blob/dc5d9ced8cc69e6f66d1c2eb55003e5720af1039/scripts/asset-source-config.js)

where this happens:

```
config.module.rules.forEach(rule => {
    rule.resourceQuery = { not: [/raw/] };
  });
  config.module.rules.push(
    {
      resourceQuery: /raw/,
      type: 'asset/source',
    });
  return config;
```

**Comment by MichaelAllenWarner** [comment]
> btw. there is another way, which would work without touching any other part of the config:
> 
> ```js
>       {
>         test: /\.txt$/,
>         type: "asset/source",
>       },
> ```
> 
> ```
> import source from "file.txt!=!./file.js";
> ```
> 
> This treats content of file.js as it was written in `file.txt`.
> 
> Not sure if this works with `require.context`, but you could try it with `require.context("code.txt!=!./code", false, /\.js$/)`, but this might look a bit weird anyway.
> 
> Otherwise you could also go with:
> 
> ```
> import sources from "./all-code.js!";
> ```
> 
> and generate code like that in the `all-code.js` **loader**:
> 
> ```js
> export { default as abc } from "code/abc.js.txt!=!./code/abc.js";
> export { default as def } from "code/def.js.txt!=!./code/def.js";
> // ...
> ```

This is extremely ugly but worked for me. Thank you!
